### PR TITLE
Splits CPU and CUDA fusion compilers

### PR DESCRIPTION
**Pull request by mruberry**
This PR splits the CPU and CUDA fusion compilers, putting them into a new jit/fusers/ directory with jit/fusers/common for common components. In particular:

- A fusion interface is created that allows "fusion handles" to be requested
- The CPU and CUDA fusers implement this interface, with dispatch determined by device
- The fusion compilers, fusion function specializations and resource strings are split
- CPU-specific classes like TempFile and DynamicLibrary are in the CPU fuser
- Common classes likes TensorDesc and the base fusion function class are in jit/fusers/common
- There is still some specialization in jit/fusers/common, but these specializations are small(-ish)
- Updates the build system to remove the dummy interface on Windows and minimize the use of macros

This structure should allow in-flight PRs to easily rebase while providing a clear interface to the fusers.

### CODE REVIEW

**Comment by mruberry** [comment]
Looks like all tests are passing. fyi @zdevito @apaszke

**Review by apaszke — COMMENTED** [review]
Looks much better than the initial PR. Some naming requests and this should be ready to land. I'd also like @ezyang to double check the use of macros, as I'm not sure if it fits well with our build philosophy.

File: torch/CMakeLists.txt
Comment (ezyang) [inline torch/CMakeLists.txt]: Yeah, this isn't going to fly. If you were only using the macros from C++ files, it'd be OK, but you've got the macros in headers, which means that you need to make sure that they get durably set, even if other projects include your header in question.

There are two ways to do this:

1. Use `target_compile_definitions`. Example in `caffe2/CMakeLists.txt`

```
  # NB: This must be target_compile_definitions, not target_compile_options,
  # as the latter is not respected by nvcc
  target_compile_definitions(caffe2 PUBLIC "AT_CORE_STATIC_WINDOWS=1")
```

This will be propagated to downstream projects IFF they use cmake to do the build, and use the installed cmake config file. This is somewhat error prone.

2. Write out a macro file, config.h style, which specifies if we built with or without the support. Example `aten/src/ATen/Config.h.in`. A quick grep suggests we don't have anything equivalent in torch at the moment.

**Review by ezyang — DISMISSED** [review]
Macro treatment is problematic.

File: torch/csrc/jit/fusers/common/common_fusion_function.h
Comment (mruberry) [reply torch/csrc/jit/fusers/common/common_fusion_function.h]: Sure.

File: torch/csrc/jit/fusers/common/common_fusion_function.h
Comment (mruberry) [reply torch/csrc/jit/fusers/common/common_fusion_function.h]: Sounds good.

File: torch/csrc/jit/fusers/common/common_fusion_handle.h
Comment (mruberry) [reply torch/csrc/jit/fusers/common/common_fusion_handle.h]: "Handle" was the term we've been using, and I think FusedKernelCache is more reflective of an implementation than interface concept. 

What about FusionHandle and FusionHandleImpl?

File: torch/csrc/jit/fusers/fusion_interface.h
Comment (mruberry) [reply torch/csrc/jit/fusers/fusion_interface.h]: Sounds good.

File: torch/CMakeLists.txt
Comment (mruberry) [reply torch/CMakeLists.txt]: Thanks Ed, this is really helpful.  

Since you're the expert, would you tell me which approach you'd prefer? I'm up for either and it sounds like you're leaning towards (1)?

File: torch/csrc/jit/fusers/common/common_fusion_handle.h
Comment (apaszke) [reply torch/csrc/jit/fusers/common/common_fusion_handle.h]: Would that mean that `FusionHandle == FusedKernelCache` and `FusionHandleImpl == FusedKernel`? I think that's ok, but I'd like us to move away from the "fusion" nomenclature, and I think kernel/function is clearer than handle which is a more generic term. What we have here is a callable.

**Review by apaszke — APPROVED** [review]
This LGTM. @ezyang can you please take a look at macros again?

**Comment by mruberry** [comment]
Naming and config changes are in, I also took the three relevant PRs that updated the fusion_compiler and ported them to the split.

I am unsure of what's going on with the two failures. The circleci appears to be running itself again (?) and pytorch-linux-trusty-py2.7.9 has had jit problems (there's a PR to disable a test for it), but in this case there are three errors, not just test_scalar_fusion.

File: torch/csrc/jit/fusers/Config.h
Comment (ezyang) [inline torch/csrc/jit/fusers/Config.h]: Don't check this in!! You need to add this file to .gitignore

File: torch/CMakeLists.txt
Comment (ezyang) [inline torch/CMakeLists.txt]: It... would be better to not dunk this in the current source dir, but I suppose it will do for now...

**Comment by ezyang** [comment]
This will need fbcode changes. @apaszke, are you going to land it? You'll need to add appropriate settings for the new Config.h file.

File: torch/csrc/jit/fusers/Config.h
Comment (mruberry) [reply torch/csrc/jit/fusers/Config.h]: Right! I did remove it but adding it to .gitignore would be safer. I'll get that fixed up.

Comment (ezyang) [reply torch/csrc/jit/fusers/Config.h]: I would have reviewed all your changes in one go but GitHub wasn't smart enough to deal with the merge commits haha

**Comment by ezyang** [comment]
@goldsborough, do we have any checks in place to make sure this new Config.h header is actually installed?

**Comment by apaszke** [comment]
I can land it, but I'll need some guidance for fbcode changes, because I know very little about its build system.

**Comment by apaszke** [comment]
@mruberry this is ready to land, but unfortunately there's a conflict in CMakeLists.txt 😕

**Comment by soumith** [comment]
conflict looks trivial when I look at the Web Editor. @apaszke Can you resolve it with "Resolve conflicts" button, or even better, on the imported diff.

**Comment by mruberry** [comment]
Thanks for landing it on your end, @apaszke. I think I resolved the conflict.

**Comment by apaszke** [comment]
@mruberry CI failure looks related

**Comment by apaszke** [comment]
Hmm that might be a transient issue. I can't reproduce this locally. Let's see if a retest fixes it.

**Comment by apaszke** [comment]
Nope, this seems to consistently fail 😕

**Comment by zou3519** [comment]
CPU fuser tests are flaky, even on master, due to some memory issues I belive

**Comment by mruberry** [comment]
It is strange that three tests were failing consistently (test_scalar_fusion was also failing until it was disabled), and this is mentioned above. I suspect these failures are revealed by this PR more than caused by it.

**Comment by apaszke** [comment]
I agree, but unfortunately we can't merge this until we resolve the problem. It's likely that we will disable the CPU fuser by default, but we don't want to disable its tests entirely to avoid undetected breakages of that code path. @zou3519 is looking into that

**Comment by zou3519** [comment]
I'm still looking into it, sorry for the delay. Some strange things are going on

**Comment by zou3519** [comment]
I posted an update in #11360 about what I've found. The summary is that test_jit.py has high peak memory usage that causes the graph fuser to fail when it runs the `fork()` syscall, despite having copy-on-write semantics.

The fastest solution to unblock this PR would be to move the fuser tests to run first in test_jit.py so they run before test_jit starts using a lot of memory. A more robust solution (that I am looking into right now) is to figure out why test_jit.py uses so much memory (> 4gb) and fix that.

**Comment by zou3519** [comment]
@mruberry @apaszke I pushed a change to make TestEndToEndHybridFrontendModels run last in test_jit.py, which should avoid the CPU fuser test failures and unblock this PR. I am investigating a more robust solution but this PR should be good to go after the CI runs

**Review by soumith — APPROVED** [review]

**Comment by apaszke** [comment]
Hmm while OSS tests are passing, this has unfortunately triggered a failure in some internal tests... I'll push some code that disables the fuser by default (except for the tests) tomorrow in the morning.

**Comment by apaszke** [comment]
I have disabled the CPU fuser by default, except for the few tests that should actually exercise it. The memory usage regression should be now addressed thanks to @zou3519, who noticed that we run `_check_trace` with grad mode enabled (which forces us to keep two graphs in memory). We still don't have a definite answer for why the peak RSS is persistent in many environments, but it looks like `free` is simply very unwilling to return memory to the OS.

**Pull request closed without merging**
